### ISSUE TITLE
Client doesn't refresh updated row properly ?

### ISSUE THREAD

**Issue by daniel1943** [rethinkdb/horizon#649]
Using latest server/client w/ react.

1: I create new row via Horizon client @ store.

2: Update it via rethinkdb admin or python console, like:

`r.db("database").table("keywords_39c8a6fb745d").get(id).update({"randomField":55})`

3: Keep react app updated:

`horizon('keywords').watch().subscribe((kw) =>
    console.log('kw', kw));`

Line above returns 2 objects, old one(unupdated) and new one(it gives it id of -1). Is this desired behavior or bug?

It works if I refresh the page. Anyway to make it show just new object without having to refresh?

**Comment by deontologician** [comment]
@daniel1943 the id field is literally `-1`? That sounds really odd. @Tryneus any idea on this?

**Comment by daniel1943** [comment]
My mistake, didn't mean the actual ID field inside created object... Console screen:

http://i.imgur.com/KK1KJD5.png

It shows new objects, but gives it -1 and keeps old one, until refreshed.

**Comment by daniel1943** [comment]
Update: problem is ID being an array - investigating further...

**Comment by daniel1943** [comment]
So I tried new fresh project and horizon client still fails to properly update if array is being used as ID... Pretty sure now that it's a bug in horizon@client, can anyone reproduce problem?

**Comment by deontologician** [comment]
@daniel1943 could you capture the traffic on the websocket (in chrome in the network tab, there's a WS filter)

**Comment by daniel1943** [comment]
From the EIO... when data changes(vtp is filed added during update):

```
4{"data":[
{"new_val":{
   "$hz_v$":0,"id":["f8d39c40-0649-49ce-9d8f-c56076fa80c7","test"],
   "keyword":"test","owner":"f8d39c40-0649-49ce-9d8f-c56076fa80c7","vtp":33},

 "old_val":{"$hz_v$":0,"id":["f8d39c40-0649-49ce-9d8f-c56076fa80c7","test"],
  "keyword":"test","owner":"f8d39c40-0649-49ce-9d8f-c56076fa80c7"},"type":    "change"}],"request_id":3}
```

Actual websocket frames:

http://i.imgur.com/7LjXv0o.png

And headers if it helps:

```
GET ws://127.0.0.1:9095/sockjs-node/494/0rx5yodp/websocket HTTP/1.1
Host: 127.0.0.1:9095
Connection: Upgrade
Pragma: no-cache
Cache-Control: no-cache
Upgrade: websocket
Origin: http://localhost:3000
Sec-WebSocket-Version: 13
User-Agent: Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/51.0.2704.106 Safari/537.36
Accept-Encoding: gzip, deflate, sdch
Accept-Language: en,en-US;q=0.8,sl;q=0.6
Cookie: csrftoken=zVAsZCI9QO1WIi2JEK0AtLWACLzibO7S
Sec-WebSocket-Key: siJDEt0OrUa0tycfeUNzRw==
Sec-WebSocket-Extensions: permessage-deflate; client_max_window_bits
```

**Comment by deontologician** [comment]
Ah, could you try out the current `next` branch? A lot of stuff has happened since then and it may have been fixed (we got rid of engine.io which was causing a lot of problems). There was also a fix with how we detect ids that are arrays if I recall correctly

**Comment by daniel1943** [comment]
Unfortunately same issue with `next` branch.

**Comment by marshall007** [comment]
@deontologician I think [this line](https://github.com/rethinkdb/horizon/blob/3a4ad8f4e02985c8a46d5d62c523861c078070b5/client/src/ast.js#L188) (and the `uninitial` case) just need a deep equality check on `id`.

**Comment by deontologician** [comment]
@marshall007 thanks. I'll fix this tomorrow, looks like 2.0 is coming out sometimes early next week, so it won't be long for @daniel1943

**Comment by deontologician** [comment]
Fix up in #656
